processing/src/main/java/org/apache/druid/segment/QueryableIndexSegment.java

```diff
@@ -19,24 +19,58 @@
 
 package org.apache.druid.segment;
 
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
 import org.apache.druid.timeline.SegmentId;
 import org.joda.time.Interval;
 
 /**
+ *
  */
 public class QueryableIndexSegment implements Segment
 {
-  private final QueryableIndex index;
-  private final QueryableIndexStorageAdapter storageAdapter;
+  private final Supplier<QueryableIndex> indexSupplier;
+  private final Supplier<QueryableIndexStorageAdapter> queryableIndexStorageAdapterSupplier;
```

Comment:
I wonder if there is a way to limit the supplier to being part of the `FireHydrant` instead of doing this modification here?

While I can't _think_ of any ill side-effect that this might cause since these suppliers shouldn't be called too many times in the scheme of things, this change also has huge surface area because of it happening here instead of being limited to ingestion.

Reply:
I rolled back the memoization code....